sxt/proof/inner_product/scalar_fold_kernel.cc

```diff
@@ -27,68 +28,37 @@
-static xena::future<> fold_scalars_case1(basct::span<s25t::element> scalars,
-                                         const s25t::element& m_low, const s25t::element& m_high,
-                                         unsigned mid, unsigned m) noexcept {
-  auto data = scalars.data();
-  // clang-format off
-  auto f = [
-    data, 
-    m_low, 
-    m_high, 
-    mid
-  ] __device__ __host__(unsigned /*m*/, unsigned i) noexcept 
-  {
-    auto& x = data[i];
-    s25o::mul(x, m_low, x);
-    s25o::muladd(x, m_high, data[mid + i], x);
-  };
-  // clang-format on
-  return algi::for_each(f, m);
-}
-
-//--------------------------------------------------------------------------------------------------
-// fold_scalars_case2
-//--------------------------------------------------------------------------------------------------
-static xena::future<> fold_scalars_case2(basct::span<s25t::element> scalars,
-                                         const s25t::element& m_low, unsigned mid,
-                                         unsigned m) noexcept {
-  auto data = scalars.data() + m;
-  // clang-format off
-  auto f = [
-    data, 
-    m_low
-  ] __device__ __host__(unsigned /*m*/, unsigned i) noexcept 
-  {
-    auto& x = data[i];
-    s25o::mul(x, m_low, x);
-  };
-  // clang-format on
-  return algi::for_each(f, mid - m);
-}
-
-//--------------------------------------------------------------------------------------------------
-// fold_scalars
-//--------------------------------------------------------------------------------------------------
-xena::future<> fold_scalars(basct::span<s25t::element> scalars, const s25t::element& m_low,
-                            const s25t::element& m_high, unsigned mid) noexcept {
-  auto n = scalars.size();
+xena::future<> async_fold_scalars(basct::span<s25t::element> scalars_p,
+                                  basct::cspan<s25t::element> scalars, const s25t::element& m_low,
+                                  const s25t::element& m_high) noexcept {
+  auto mid = scalars_p.size();
   SXT_DEBUG_ASSERT(
       // clang-format off
-      basdv::is_active_device_pointer(scalars.data()) &&
-      0 < mid && 
-      mid < n &&
-      n <= 2u * mid
+      scalars_p.size() == mid &&
+      mid < scalars.size() && scalars.size() <= 2u * mid
       // clang-format on
   );
-  auto m = n - mid;
-  auto fut1 = fold_scalars_case1(scalars, m_low, m_high, mid, m);
-  if (m == mid) {
-    return fut1;
-  }
-  auto fut2 = fold_scalars_case2(scalars, m_low, mid, m);
-  return xena::await_all(std::move(fut1), std::move(fut2));
+  auto f1 = [m_low, m_high] __device__ __host__(s25t::element & x,
+                                                const s25t::element& y) noexcept {
+    s25o::mul(x, m_low, x);
+    s25o::muladd(x, m_high, y, x);
+  };
+  auto m = scalars.size() - mid;
+  basit::chunk_options chunk_options{
+      .min_size = 2u << 10u,
+      .max_size = 2u << 20u,
+  };
```

Comment:
I wonder if a comment would help here too? Only if the chuck options need to be tested and optimized.

Reply:
Added commenting.